### PR TITLE
feat: support machinepools in bootstrapper

### PR DESCRIPTION
**Pull request by rsmitty**
This PR adds support for the machinepool feature gate. When enabled, we
can now catch new machinepools, generate userdata, and populate the
secret as expected.

This is necessary for using Autoscaling Groups in AWS.

Signed-off-by: Spencer Smith <robertspencersmith@gmail.com>

### CODE REVIEW

File: controllers/secrets.go [hunk 1]
```diff
@@ -121,12 +121,16 @@ func (r *TalosConfigReconciler) writeK8sCASecret(ctx context.Context, scope *Tal
...
+
+	r.Log.Info("handling bootstrap data for ", "owner", ownerName)
+
+	_, err := r.fetchSecret(ctx, scope.Config, scope.ConfigOwner.GetName()+"-bootstrap-data")
 	if k8serrors.IsNotFound(err) {
 		certSecret := &corev1.Secret{
 			ObjectMeta: metav1.ObjectMeta{
 				Namespace: scope.Config.Namespace,
-				Name:      scope.Machine.Name + "-bootstrap-data",
+				Name:      scope.ConfigOwner.GetName() + "-bootstrap-data",
```
Comment (smira) [inline controllers/secrets.go:133]: nit: we might use `ownerName` here

File: controllers/talosconfig_controller.go [hunk 2]
```diff
@@ -214,17 +255,31 @@ func (r *TalosConfigReconciler) Reconcile(req ctrl.Request) (_ ctrl.Result, rerr
 	}
 
 	// Packet acts a fool if you don't prepend #!talos to the userdata
```
Comment (smira) [inline controllers/talosconfig_controller.go:257]: I guess this comment should be moved down close to `PacketMachine` now

File: controllers/talosconfig_controller.go [hunk 3]
```diff
@@ -80,16 +86,51 @@ type talosConfigContext struct {
...
+	r.Scheme = mgr.GetScheme()
+
+	b := ctrl.NewControllerManagedBy(mgr).
 		For(&bootstrapv1alpha3.TalosConfig{}).
-		Complete(r)
+		WithOptions(options).
+		Watches(
+			&source.Kind{Type: &capiv1.Machine{}},
+			&handler.EnqueueRequestsFromMapFunc{
+				ToRequests: handler.ToRequestsFunc(r.MachineToBootstrapMapFunc),
```
Comment (smira) [inline controllers/talosconfig_controller.go:98]: dumb question: how did our provider work before? if it didn't have this watch?

**Review by smira — COMMENTED** [review]
some random nits

File: controllers/talosconfig_controller.go (hunk 3 above)
Comment (rsmitty) [reply controllers/talosconfig_controller.go:98]: If i understand it correctly myself, this only comes into play when you add the extra "watches" to the controller. By default, it owns "TalosConfig" so would only trigger a reconcile when that resource changed. But adding watches for Machines and MachinePools requires the handler functions. 

We also do something similar in the controlplane provider where we have extra watchers for cluster resource.

File: controllers/secrets.go (hunk 1 above)
Comment (rsmitty) [reply controllers/secrets.go]: yep, good call. thx.

File: controllers/talosconfig_controller.go (hunk 2 above)
Comment (rsmitty) [reply controllers/talosconfig_controller.go:257]: That whole if/else block below relates to PacketMachine, as we have to handle patching MachinePool and Machines separately.

File: controllers/talosconfig_controller.go (hunk 3 above)
Comment (smira) [reply controllers/talosconfig_controller.go:98]: yep, I understand that, I mean if watch for Machine is required for the provider to function properly, how did it work before this PR? as it doesn't watch Machine?

File: controllers/talosconfig_controller.go (hunk 2 above)
Comment (smira) [reply controllers/talosconfig_controller.go:257]: ah, sorry, missed that

**Review by smira — APPROVED** [review]
I think we need to seriously consider adding testbed for these providers, be it Sfyra, some mocking or anything else

**Comment by rsmitty** [comment]
100% agreed. I'll add a ticket to backlog to track that.

**Pull request merged**